README.md
# Distribute

**Distribute** is a middleware-based API to expressively perform request
routing / load balancing in Node.JS.

## How to use

### Normal HTTP

```js
var http = require('http').createServer().listen(3000)
  , srv = require('distribute')(http);

srv.use(function (req, res, next) {
  if (req.headers.host == 'blog.mydomain.com') {
    next(8000);
  } else {
    next();
  }
});

srv.use(function (req, res, next) {
  somethingAsync(function (err, host, port) {
    if (err) return next(err); // sends a `500` and cleans up
    next(port, host);
  });
});
```

### WebSocket

Requests triggered by the `upgrade` event (as a result of the `Upgrade`
HTTP header) are handled by prepending the `ws` flag each time you call
`use`.

```js
server.ws.use(function (req, socket, next) {
  next(3000);
});
```

## Features

- Leverages the well-tested `node-http-proxy`.
- Simplicity of Express.
- Compatible with connect middleware (eg: qs parser, cookie decoder).
- Middleware makes sticky/session load balancing trivial to write.
- Middleware can perform async tasks.
  **Distribute** manages buffers transparently for you.

## API

### next

The `next` parameter can take three signatures:

  - no parameters (`next()`) will execute the next middleware. If no
    middleware is available, an error is displayed (refer to the "Default
    behaviors section")
  - port (`Number`)
  - port (`Number`), host (`String`)

### req

**Distribute** adds two properties to request objects:

#### req.buf

The `node-http-proxy` data buffer.

#### req.head

For WS requests, the first packet of the stream, only present for legacy
purposes.

## Behaviors

### Error handling

When an `Error` object is passed to next, or when no middleware will
handle a given request, the default behavior is to show a
`500 Internal Server Error` (for HTTP requests) or the socket is ended 
(WS requests). In development (ie: NODE_ENV is set to `development`), a
stack trace is sent along with the error code.

If you want to define custom "error handling middleware", you can do so
by adding a function with 4 parameters instead of 3 (in other words, with
an arity of 4).

```js
// regular requests
srv.use(function (err, req, res, next) {
  next();
});

// ws requests
srv.use(function (err, req, socket, next) {
  next();
});
```

It's not necessary to pass the error to `next` to trigger the next error 
middleware.

### Buffers

Request data buffers are cleaned up automatically when:

  - a response is produced prematurely instead of proxying. For example

  ```js
  srv.use(function (req, res, next) {
    res.writeHead(204);
    res.end();
  });
  ```

  - a socket for an upgrade is `.end` or `.destroy` prematurely:

  ```js
  srv.ws.use(function (req, socket, next) {
    socket.end();
  });
  ```

## License 

(The MIT License)

Copyright (c) 2011 Guillermo Rauch &lt;guillermo@learnboost.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

0.1.4 / 2012-02-27
==================

  * Added guard for double next.

0.1.3 / 2012-02-03
==================

  * Added reference to request from response obj.
  * Added `req.next`, `res.next`, `err.next`.
  * Added `req.res` reference.
  * Fixed `ws` getter.

0.1.2 / 2012-01-30
==================

  * Fix for websocket proxying (fixes #2).

0.1.1 / 2012-01-26
==================

  * Middleware scope is now the `Distributor` instance.
  * Fixed default WS middleware.
  * Fixed WS proxying (thanks @indexzero).

0.1.0 / 20012-01-18
===================

  * Initial release


test/distribute.js

/**
 * Test dependencies.
 */

var http = require('http')
  , expect = require('expect.js')
  , request = require('superagent')
  , distribute = require('../distribute')

/**
 * Test.
 */

describe('distribute', function () {

  it('should support multiple request middleware', function (done) {
    var servers = 3;

    var srv1 = http.createServer(function (req, res) {
      res.writeHead(200);
      res.end('a');
    }).listen(4001, onListen);

    var srv2 = http.createServer(function (req, res) {
      res.writeHead(200);
      res.end('b');
    }).listen(4002, onListen);

    var master = http.createServer().listen(4000, onListen);
    var srv = distribute(master)

    srv.use(function (req, res, next) {
      if (/a\./.test(req.headers.host)) return next(4001);
      next();
    });

    srv.use(function (req, res, next) {
      if (/b\./.test(req.headers.host)) return next(4002);
      next();
    });

    function onListen () {
      if (!--servers) {
        var total = 2;

        function finish () {
          var total = 3;

          function onClose () {
            --total || done();
          }

          srv1.once('close', onClose).close();
          srv2.once('close', onClose).close();
          master.once('close', onClose).close();
        }

        request.get('http://localhost:4000')
          .set('Host', 'a.localhost')
          .end(function (res) {
            expect(res.text).to.be('a');
            --total || finish();
          });

        request.get('http://localhost:4000')
          .set('Host', 'b.localhost')
          .end(function (res) {
            expect(res.text).to.be('b');
            --total || finish();
          });
      }
    }
  });

  it('should support middleware chaining', function (done) {
    var servers = 3;

    var srv1 = http.createServer(function (req, res) {
      res.writeHead(200);
      res.end('a');
    }).listen(4001, onListen);

    var srv2 = http.createServer(function (req, res) {
      res.writeHead(200);
      res.end('b');
    }).listen(4002, onListen);

    var master = http.createServer().listen(4000, onListen);
    var srv = distribute(master)
      .use(function (req, res, next) {
        if (/a\./.test(req.headers.host)) return next(4001);
        next();
      })
      .use(function (req, res, next) {
        if (/b\./.test(req.headers.host)) return next(4002);
        next();
      })

    function onListen () {
      if (!--servers) {
        var total = 2;

        function finish () {
          var total = 3;

          function onClose () {
            --total || done();
          }

          srv1.once('close', onClose).close();
          srv2.once('close', onClose).close();
          master.once('close', onClose).close();
        }

        request.get('http://localhost:4000')
          .set('Host', 'a.localhost')
          .end(function (res) {
            expect(res.text).to.be('a');
            --total || finish();
          });

        request.get('http://localhost:4000')
          .set('Host', 'b.localhost')
          .end(function (res) {
            expect(res.text).to.be('b');
            --total || finish();
          });
      }
    }
  });

  it('should return 501 if no middleware is left', function (done) {
    var servers = 2;
    var srv1 = http.createServer(function (req, res) {
      res.writeHead(200);
      res.end('a');
    }).listen(4001, onListen);
    var master = http.createServer().listen(4000, onListen);
    var srv = distribute(master)

    srv.use(function (req, res, next) {
      if (/a\./.test(req.headers.host)) return next(4001);
      next();
    });

    function onListen () {
      if (!--servers) {
        var total = 2;

        function finish () {
          var total = 2;

          function onClose () {
            --total || done();
          }

          srv1.once('close', onClose).close();
          master.once('close', onClose).close();
        }

        request.get('http://localhost:4000')
          .set('Host', 'a.localhost')
          .end(function (res) {
            expect(res.text).to.be('a');
            --total || finish();
          });

        request.get('http://localhost:4000')
          .set('Host', 'b.localhost')
          .end(function (res) {
            expect(res.status).to.be(501);
            --total || finish();
          });
      }
    }
  });

  it('should do error handling', function (done) {
    var servers = 2;
    var srv1 = http.createServer(function (req, res) {
      res.writeHead(200);
      res.end('a');
    }).listen(4001, onListen);
    var master = http.createServer().listen(4000, onListen);
    var srv = distribute(master)

    srv.use(function (req, res, next) {
      if (/a\./.test(req.headers.host)) return next(4001);
      next(new Error('Test'));
    });

    srv.use(function (req, res, next) {
      throw new Error('should not execute');
    });

    srv.use(function (err, req, res, next) {
      expect(err).to.be.an(Error);
      expect(err.message).to.be('Test');
      next();
    });

    function onListen () {
      if (!--servers) {
        var total = 2;

        function finish () {
          var total = 2;

          function onClose () {
            --total || done();
          }

          srv1.once('close', onClose).close();
          master.once('close', onClose).close();
        }

        request.get('http://localhost:4000')
          .set('Host', 'a.localhost')
          .end(function (res) {
            expect(res.text).to.be('a');
            --total || finish();
          });

        request.get('http://localhost:4000')
          .set('Host', 'b.localhost')
          .end(function (res) {
            expect(res.status).to.be(500);
            if ('development' == process.env.NODE_ENV) {
              expect(res.text).to.contain('Test');
            }
            --total || finish();
          });
      }
    }
  });

  it('should buffer', function (done) {
    var srv1 = http.createServer(function (req, res) {
      expect(req.method).to.be('POST');

      var data = '';

      req.on('data', function (chunk) {
        data += chunk;
      });

      req.on('end', function () {
        expect(data).to.be('ABC');
        res.end();
        srv1.close();
        master.on('close', done).close();
      });
    }).listen(4001, onListen);

    var master = http.createServer().listen(4000, onListen);
    var chunks = 0;
    var srv = distribute(master)
      .use(function (req, res, next) {
        expect(req.headers.host).to.be('a.localhost');
        expect(chunks).to.be(1);

        setTimeout(function () {
          next(4001);
        }, 100);
      });

    var total = 2;
    function onListen () {
      if (--total) return;
      var req = request
        .post('http://localhost:4000')
        .set('Host', 'a.localhost')
        .set('Content-Type', 'text/plain');

      req.write('A');

      chunks++;
      setTimeout(function () {
        chunks++;
        req.write('B');

        setTimeout(function () {
          chunks++;
          req.write('C');
          req.end();
        }, 20);
      }, 20);
    }
  });

  it('should not break on double next', function (done) {
    var httpServer = http.createServer()
      , srv = distribute(httpServer)

    srv.use(function (req, res, next) {
      next();
      setTimeout(function () {
        expect(next).to.throwException(/more than once.*tobi/);
      }, 10);
    });

    srv.use(function (req, res, next) {
      setTimeout(function () {
        res.writeHead(200);
        res.end('Hello World');
      }, 50);
    });

    httpServer.listen(4000, function () {
      request.get('http://localhost:4000/tobi').end(function (res) {
        expect(res.text).to.be('Hello World');
        httpServer.on('close', done).close();
      });
    });
  });

});


distribute.js

/**
 * Module dependencies.
 */

var proxy = require('http-proxy')
  , RoutingProxy = proxy.RoutingProxy
  , EventEmitter = require('events').EventEmitter
  , debug = require('debug')('distributor')

/**
 * Module exports.
 */

module.exports = exports = Distributor;

/**
 * Version.
 *
 * @api public
 */

exports.version = '0.1.4';

/**
 * Distributor factory/constructor.
 *
 * @param {http.Server} server that acts as balancer
 * @api public
 */

function Distributor (server) {
  if (global == this) return new Distributor(server);

  this.server = server;
  this.proxy = new RoutingProxy;
  this.middleware = {
      ws: { all: [this.defaultWS], error: [this.defaultWSError] }
    , http: { all: [this.defaultHTTP], error: [this.defaultHTTPError] }
  };

  var self = this;
  server.on('request', function (req, res) {
    self.onRequest(req, res);
  });
  server.on('upgrade', function (req, socket, head) {
    self.onUpgrade(req, socket, head);
  });
}

/**
 * Inherits from EventEmitter.
 */

Distributor.prototype.__proto__ = EventEmitter.prototype;

/**
 * Captures the `ws` namespace and sets the state.
 */

Distributor.prototype.__defineGetter__('ws', function () {
  this.useWS = true;
  return this;
});

/**
 * Appends a middleware.
 *
 * @return {Distributor} for chaining
 * @api public
 */

Distributor.prototype.use = function (fn) {
  var stack = this.middleware;

  if (this.useWS) {
    stack = stack.ws;
  } else {
    stack = stack.http;
  }

  if (4 == fn.length) {
    stack = stack.error;
  } else {
    stack = stack.all;
  }

  // make sure we keep the default middlewares last
  stack.splice(stack.length - 1, 0, fn);

  this.useWS = false;

  return this;
};

/**
 * Prepares a request for middleware execution.
 *
 * @param {http.ServerRequest} incoming request
 * @return {http.ServerRequest} processed request
 * @api private
 */

Distributor.prototype.prepare = function (req) {
  req.buf = this.proxy.buffer(req);
  return req;
};

/**
 * Handles regular http requests.
 *
 * @api private
 */

Distributor.prototype.onRequest = function (req, res) {
  var self = this;

  // set up patches
  res.req = req;
  req.res = res;
  req.buf = proxy.buffer(req);

  // set up buffering cleanup
  function onFinish () {
    req.buf.destroy();
  }
  res.on('finish', onFinish);

  var stack = this.middleware.http;

  this.run(stack.all, [req, res], function (port, host) {
    if (port instanceof Error) {
      self.run(stack.error, [port, req, res]);
    } else {
      res.removeListener('finish', onFinish);
      var proxy = { port: port, host: host || 'localhost', buffer: req.buf };
      self.proxy.proxyRequest(req, res, proxy);
    }
  });
};

/**
 * Handles http upgrades.
 *
 * @api private
 */

Distributor.prototype.onUpgrade = function (req, socket, head) {
  var self = this;

  // set up buffering
  req.buf = proxy.buffer(req);

  // append legacy `.head`
  req.head = head;

  // set up buffering cleanup
  function onClose () {
    req.buf.destroy();
  }
  socket.on('close', onClose);

  var stack = this.middleware.ws;

  self.run(stack.all, [req, socket], function (port, host) {
    if (port instanceof Error) {
      debug('running error middleware for ws request');
      self.run(stack.error, [port, req, res]);
    } else {
      debug('proxying ws request');
      socket.removeListener('close', onClose);
      var proxy = { port: port, host: host || 'localhost', buffer: req.buf };
      self.proxy.proxyWebSocketRequest(req, socket, req.head, proxy);
    }
  });
};

/**
 * Runs a stack of middleware.
 *
 * @param {Array} stack of functions
 * @param {Array} parameters to pass to middleware
 * @param {Function} called when a middleware calls next with parameters
 * @api private
 */

Distributor.prototype.run = function (stack, params, onNext) {
  var l = params.length
    , self = this

  function step (i) {
    var called = false;

    // set up `next` as last parameter
    params[l] = function next () {
      if (called) {
        var req = params[0];
        throw new Error('`next` called more than once in\n' + stack[i]
          + ' for ' + req.url + ' (' + JSON.stringify(req.headers) + ')');
      } else {
        called = true;
        if (arguments.length) return onNext.apply(self, arguments);
        step(++i);
      }
    };

    // add `.next` to all objects in the parameters except last one
    for (var ii = 0; ii < l; ii++) {
      params[ii].next = params[l];
    }

    stack[i].apply(self, params);
  }

  step(0);
};

/**
 * Default handler middleware.
 *
 * @api private
 */

Distributor.prototype.defaultHTTP = function (req, res, next) {
  debug('executing default request middleware');
  res.writeHead(501);
  res.end();
};

/**
 * Default error handler middleware.
 *
 * @api private
 */

Distributor.prototype.defaultHTTPError = function (err, req, res, next) {
  debug('executing default request error middleware');

  if ('development' == process.env.NODE_ENV) {
    res.writeHead(500, {
        'Content-Type': 'text/plain'
      , 'Content-Length': Buffer.byteLength(err.stack)
    });
    res.end(err.stack);
  } else {
    res.writeHead(500);
    res.end();
  }
};

/**
 * Default ws handler middleware
 *
 * @api upgrade
 */

Distributor.prototype.defaultWS = function (req, socket, next) {
  debug('executing default ws middleware');
  socket.end();
};

/**
 * Default ws error handler middleware
 *
 * @api upgrade
 */

Distributor.prototype.defaultWSError = function (err, req, socket, next) {
  debug('executing default ws error middleware');
  socket.end();
};
